src/main/java/htsjdk/samtools/cram/structure/Container.java

```diff
@@ -17,168 +17,363 @@
+        final AlignmentContext alignmentContext = getDerivedAlignmentContext(commonRefContext);
+
+        int baseCount = 0;
+        int blockCount = 0;
+        int recordCount = 0;
+        for (final Slice slice : slices) {
+            recordCount += slice.getNumberOfRecords();
+            blockCount += slice.getNumberOfBlocks();
+            baseCount += slice.getBaseCount();
+        }
 
-    // minimum alignment start of the reads in this Container
-    // uses a 1-based coordinate system
-    public int alignmentStart = Slice.NO_ALIGNMENT_START;
-    public int alignmentSpan = Slice.NO_ALIGNMENT_SPAN;
-    public int nofRecords = 0;
-    public long globalRecordCounter = 0;
+        this.containerHeader = new ContainerHeader(
+                alignmentContext,
+                globalRecordCounter,
+                blockCount,
+                recordCount,
+                baseCount);
 
-    public long bases = 0;
-    public int blockCount = -1;
+        validateContainerReferenceContext();
+        checkReferenceContexts(commonRefContext.getReferenceContextID());
+    }
 
     /**
-     * Slice byte boundaries as offsets within this container,
-     * counted after the container header.  The start of the compression header
-     * has offset 0.
+     * Create a container for use by CramContainerHeaderIterator, which is only used to
+     * find the offsets within a CRAM stream where containers start.
      *
-     * Equal to {@link Slice#byteOffsetFromCompressionHeaderStart}.
-     *
-     * As an example, suppose we have:
-     * - landmarks[0] = 9000
-     * - landmarks[1] = 109000
-     * - containerBlocksByteSize = 123456
+     * @param containerHeader the container header for the container
+     * @param containerByteOffset the byte offset of this container in the containing stream
+     */
+    public Container(final ContainerHeader containerHeader, final long containerByteOffset) {
+        this.containerHeader = containerHeader;
+        this.containerByteOffset = containerByteOffset;
+        compressionHeader = null;
+        slices = Collections.EMPTY_LIST;
+    }
+
+    /**
+     * Read a Container from a CRAM stream. This reads the container header and all slice blocks, but does
+     * not resolve the blocks into CRAMRecords, since we don't want to do that until its necessary (and it
+     * might not be if, for example, we're indexing the container).
      *
-     * We therefore know:
-     * - the compression header size = 9000
-     * - Slice 0 has offset 9000 and size 100000 (109000 - 9000)
-     * - Slice 1 has offset 109000 and size 14456 (123456 - 109000)
+     * @param version CRAM version of the input stream
+     * @param inputStream input stream to read
+     * @param containerByteOffset byte offset within the stream of this container
      */
-    public int[] landmarks;
+    public Container(final Version version, final InputStream inputStream, final long containerByteOffset) {
```

Comment:
Container/Slice/CompressionHeader are all resolved in favor of constructors to keep them as immutable as possible. Block and SliceBlocks have read methods because they need to be mutated after creation.